### PR TITLE
Fix the NPE when a worker is defined after an if block without an else block

### LINKED ISSUE #35765: NPE when a worker is defined after an if block without an else block
In current master(bce5c1698ee823d76250727e30af721836a9659f) building websub module throws this exception.

```
[2022-04-09 13:22:07,310] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.resolveAssociatedWorkerFunc(SymbolEnter.java:4547)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.createVarSymbol(SymbolEnter.java:4529)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.createVarSymbol(SymbolEnter.java:4515)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineVarSymbol(SymbolEnter.java:4489)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:2294)
        at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:54)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:300)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.handleDeclaredWithVar(SemanticAnalyzer.java:1622)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:959)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:59)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4101)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4077)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4058)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1926)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:55)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4101)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4077)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4062)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1920)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:70)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4101)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4077)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4062)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:511)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4101)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:486)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:81)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4101)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4077)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4058)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:323)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$2(SemanticAnalyzer.java:335)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:335)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:235)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:172)
        at org.wso2.ballerinalang.compiler.tree.SimpleBLangNodeAnalyzer.visitNode(SimpleBLangNodeAnalyzer.java:201)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:282)
        at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:188)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:436)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:386)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:213)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:198)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:114)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:109)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:233)
        at io.ballerina.projects.Package.getCompilation(Package.java:147)
        at io.ballerina.projects.Package.runCodeGenAndModifyPlugins(Package.java:254)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:95)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.PackCommand.execute(PackCommand.java:247)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [websocket:(1:1,1:1)] Compilation failed due to an unhandled exception
error: compilation contains errors

```

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #35765

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -4543,10 +4542,18 @@ private void resolveAssociatedWorkerFunc(BWorkerSymbol worker, SymbolEnv env) {
 
         for (BLangLambdaFunction lambdaFn : env.enclPkg.lambdaFunctions) {
             LineRange workerBodyPos = lambdaFn.function.pos.lineRange();
+            Location targetRangePos = env.node.pos;
+
+            // TODO: targetRangePos is null, because we create a block stmt to group the statement after a If block
+            //  without en else block. Its pos is not set. Setting the pos requires the exact positions of start
+            //  and end after the if block. When the pos is set we can remove this check
+            if (targetRangePos == null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:4550]: Does this fix the issue? Still seems to crash for me.

**Review by MaryamZi — COMMENTED** [review]
Shall we update the title to convey what exactly we are fixing?

**Comment by MaryamZi** [comment]
There seem to be unrelated commits also?

**Comment by gimantha** [comment]
> There seem to be unrelated commits also?

they were reverted

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 1 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:4551]: `env.node.parent` was replaced with `env.enclInvokable`

**Review by pcnfernando — APPROVED** [review]
Approving as a temp fix

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/narrowing_with_if_without_else_not_completed_normally_test.bal [hunk 2]
```diff
@@ -428,3 +428,20 @@ function testTypeNarrowingWithEqualityInCondition2() {
 
     int j = i;
 }
+
+public function testReachabilityWithWorker1() returns error? {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/narrowing_with_if_without_else_not_completed_normally_test.bal:432]:
```suggestion
public function testTypeOfCapturedTypeNarrowedVariableInWorker() returns error? {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 3]
```diff
@@ -4543,10 +4543,18 @@ private void resolveAssociatedWorkerFunc(BWorkerSymbol worker, SymbolEnv env) {
 
         for (BLangLambdaFunction lambdaFn : env.enclPkg.lambdaFunctions) {
             LineRange workerBodyPos = lambdaFn.function.pos.lineRange();
+            Location targetRangePos = env.node.pos;
+
+            // TODO: targetRangePos is null, because we create a block stmt to group the statement after a If block
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:4548]: Can you please create an issue to fix this properly?

**Comment by MaryamZi** [comment]
> > There seem to be unrelated commits also?
> 
> they were reverted

Ideally we shouldn't have those commits at all? 

Will merge this now anyway since it is fixing a blocker.

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/narrowing_with_if_without_else_not_completed_normally_test.bal (hunk 2 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/narrowing_with_if_without_else_not_completed_normally_test.bal:432]: Actually, this happens even without the return statement, right? So doesn't quite have anything to do with not returning normally.

I think we should move this to a separate test (worker tests maybe or if we have separate tests specifically for adding the block statement for if blocks without else blocks - cc @lasinicl) and change the name accordingly.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/narrowing_with_if_without_else_not_completed_normally_test.bal:432]: Let's also add a test like

```ballerina
function fn() {
    boolean b = true;
    
    if b {
        
    }

    worker w1 {

    }
}
```

**Pull request merged**